File: internal/cmds/fed.go
package cmds

import (
	"context"
	"database/sql"
	"fmt"
	"log"
	"strings"

	twitch "github.com/gempir/go-twitch-irc/v3"
	"github.com/pocketbase/dbx"

	"abobus/internal/services"
)

type FedCmd struct{}

func (FedCmd) Command() string {
	return "?fed"
}

func (FedCmd) Description() string {
	return "Show how many times word has been used"
}

func (cmd FedCmd) Run(ctx context.Context, s *services.Services, perms []string, message twitch.PrivateMessage) (string, error) {
	words := strings.Split(message.Message, " ")
	if len(words) != 2 && len(words) != 3 {
		return fmt.Sprintf(`Usage: "%[1]s <word>" or "%[1]s <user> <word>" or "%[1]s * <word>"`, cmd.Command()), nil
	}

	db := s.Backend.DB()

	if len(words) == 3 && words[1] == "*" {
		theWord := words[2]
		var count int
		if err := db.
			Select("COUNT(*) AS count").
			From("messages").
			Where(dbx.And(
				dbx.NewExp("channel={:channel}", dbx.Params{"channel": message.Channel}),
				// TODO: fix theWord = '%%' escaping
				dbx.Like("message", theWord),
				dbx.NotLike("message", cmd.Command()).Match(false, true),
			)).
			Row(&count); err != nil {
			if err != sql.ErrNoRows {
				// TODO: save log
				log.Println(err.Error())
				return "", err
			}
		}

		return fmt.Sprintf("Слово %s было использовано %d раз.", theWord, count), nil
	}

	var (
		theWord string
		user    string
		mention string
	)
	if len(words) == 2 {
		user = message.User.Name
		theWord = words[1]
		mention = "Ты"
	} else {
		user = words[1]
		theWord = words[2]
		mention = user
	}

	var count int
	err := db.Select("COUNT(*) AS count").From("messages").Where(dbx.And(
		dbx.NewExp("channel={:channel}", dbx.Params{"channel": message.Channel}),
		// TODO: fix theWord = '%%' escaping
		dbx.Like("message", theWord),
		dbx.NotLike("message", cmd.Command()).Match(false, true),
		dbx.NewExp("user_name={:user_name}", dbx.Params{"user_name": user}),
	)).Row(&count)
	if err != nil && err != sql.ErrNoRows {
		// TODO: save log
		log.Println(err.Error())
		return "", err
	}

	return fmt.Sprintf("%s использовал слово %s %d раз.", mention, theWord, count), nil
}

File: internal/cmds/fed.ts
import {type Command} from "../cmds.ts";
import * as message from "../message.ts";
import * as permissions from "../permissions.ts";
import * as services from "../services.ts";

export default {
  RequiredPermissions: ["execute_commands"],
  Command: "?fed",
  Description: "Show how many times word has been used",
  Run(s: services.Services, perms: permissions.PermissionsList, msg: message.TwitchMessage): string {
    const words = msg.Text.split(" ");
    if (words.length !== 2 && words.length !== 3) {
      return `Usage: "${this.Command} <word>" or "${this.Command} <user> <word>" or "${this.Command} * <word>"`;
    }

    const db = s.Backend.DB();

    if (words.length === 3 && words[1] === "*") {
      const theWord = words[2];
      const count: number = db.
        Select("COUNT(*) AS count").
        From("messages").
        Where(dbx.And(
          dbx.NewExp("channel={:channel}", {"channel": msg.Channel}),
          // TODO: fix theWord = '%%' escaping
          dbx.Like("message", theWord),
          dbx.NotLike("message", this.Command).Match(false, true),
        )).
        Row();
      return `Слово ${theWord} было использовано ${count} раз.`;
    }

    const [
      theWord,
      user,
      mention,
    ] = ((): [theWord: string, user: string, mention: string] => {
      if (words.length === 2) {
        return [words[1], msg.User.Name, "Ты"];
      } else {
        const user = words[1];
        return [words[2], user, user];
      }
    })();

    const count: number = db.Select("COUNT(*) AS count").From("messages").Where(dbx.And(
      dbx.NewExp("channel={:channel}", {"channel": msg.Channel}),
      // TODO: fix theWord = '%%' escaping
      dbx.Like("message", theWord),
      dbx.NotLike("message", this.Command).Match(false, true),
      dbx.NewExp("user_name={:user_name}", {"user_name": user}),
    )).Row()

    return `${mention} использовал слово ${theWord} ${count} раз.`;
  },
} as Command;
